Retry get requests

brettelliot opened the pull request:
Use requests.retry for all get requests to retry when the server responds with possibly recoverable status codes. Use a simple for loop to retry requests that are OK but no data was recieved.

<!-- Korbit AI PR Description Start -->
## Description by Korbit AI

### What change is being made?
Implement retry logic for GET requests in the Tradier API client to increase robustness and handle transient network errors.

### Why are these changes being made?
These changes ensure that GET requests can automatically retry on failure due to specific HTTP status codes or empty responses, reducing the likelihood of failed API interactions and improving reliability. This approach minimizes potential disruptions in services reliant on Tradier API data due to temporary network issues. The retry logic focuses on idempotent GET requests to avoid unintended side effects.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`
<!-- Korbit AI PR Description End -->

brettelliot committed:
updated readme with correct steps; added checks for .env file for testing

brettelliot committed:
retry request session for gets

brettelliot committed:
retry the request in get_historical_quotes also

brettelliot committed:
docs

brettelliot committed:
adding logging when no data retrieved for get_historical_quotes and retrying

brettelliot committed:
add retry for all gets

Manually retry tradier response succeeds but no data is returned

brettelliot committed:
cancel orders after test

cancel orders after test

brettelliot committed:
moved retry loops to request

moved retry loops to request call cleaning up code

brettelliot committed:
retry in request loop

updated how i do retries in request loop

brettelliot committed:
bump version. fix test

Im not sure why the tests failed. but when i looked at them, i wasn't sure how they were passing! Unless there's some weirdness and i ran them on a expiration date? Gonna commit this and just run the tests every day for a week and see.

davidlatte commented on lumiwealth_tradier/base.py:
```diff
@@ -77,25 +108,43 @@ def request(self, endpoint, params=None, headers=None, data=None, method="get")
+                    break
+                else:
+                    if retry_attempt == retry_attempts - 1:
+                        logger.info(f"Response from {endpoint} did not contain {required_response_key}.")
```
This should raise a TradierApiError exception because it still will not return the actual data the user is requesting.

brettelliot replied:
Done.

brettelliot replied:
Actually, now that i that, a lot of tests fail because each get call has its own specific checks for the response data and raises LookupErrors if it can't find the data it needs. I did change the logger to error so that there is a logger message before the exception is raised.

davidlatte commented on lumiwealth_tradier/base.py:
```diff
@@ -77,25 +108,43 @@ def request(self, endpoint, params=None, headers=None, data=None, method="get")
+                        logger.info(f"Response from {endpoint} did not contain {required_response_key}.")
+                    else:
+                        logger.info(f"Response from {endpoint} did not contain {required_response_key}. Retrying...")
+                        sleep(1)
```
I agree with the Bot suggestion!

brettelliot replied:
Done.

davidlatte reviewed (approved):
Overall looks good, just two tiny suggestions.

brettelliot commented on lumiwealth_tradier/base.py:
```diff
@@ -77,25 +108,43 @@ def request(self, endpoint, params=None, headers=None, data=None, method="get")
+            try:
+                ret_data = r.json()
+            except json.decoder.JSONDecodeError:
+                ret_data = {}
```
Done.

brettelliot commented on lumiwealth_tradier/base.py:
```diff
@@ -115,3 +164,29 @@ def send(self, endpoint, data, headers=None) -> dict:
         """
         return self.request(endpoint, params={}, headers=headers, data=data, method="post")
+
+    def requests_retry_session(
+            self,
+            retries=DEFAULT_RETRY_ATTEMPTS,
+            backoff_factor=DEFAULT_RETRY_BACKOFF_FACTOR,
+            status_forcelist=None,
+            allowed_methods=DEFAULT_ALLOWED_METHODS,
+            session=None,
+    ):
```
Done.

brettelliot committed:
Address comments from Dave and Korbit

* added timeout to get requests
*adding logging when we can't decode the json or when we expected data but didnt get any
